pilot/pkg/networking/core/v1alpha3/httproute.go

```diff
@@ -42,23 +40,41 @@ const wildcardDomainPrefix = "*."
 
 // BuildHTTPRoutes produces a list of routes for the proxy
 func (configgen *ConfigGeneratorImpl) BuildHTTPRoutes(env *model.Environment, node *model.Proxy, push *model.PushContext,
-	routeName string) *xdsapi.RouteConfiguration {
-	var rc *xdsapi.RouteConfiguration
+	routeNames []string) []*xdsapi.RouteConfiguration {
+	routeConfigurations := make([]*xdsapi.RouteConfiguration, 0)
+
 	switch node.Type {
 	case model.SidecarProxy:
-		rc = configgen.buildSidecarOutboundHTTPRouteConfig(env, node, push, routeName)
-		if rc != nil {
-			rc = envoyfilter.ApplyRouteConfigurationPatches(networking.EnvoyFilter_SIDECAR_OUTBOUND, node, push, rc)
+		vHostCache := make(map[int][]*route.VirtualHost)
+		for _, routeName := range routeNames {
+			rc := configgen.buildSidecarOutboundHTTPRouteConfig(env, node, push, routeName, vHostCache)
+			if rc != nil {
+				rc = envoyfilter.ApplyRouteConfigurationPatches(networking.EnvoyFilter_SIDECAR_OUTBOUND, node, push, rc)
+			} else {
+				rc = &xdsapi.RouteConfiguration{
+					Name:             routeName,
+					VirtualHosts:     []*route.VirtualHost{},
+					ValidateClusters: proto.BoolFalse,
+				}
+			}
+			routeConfigurations = append(routeConfigurations, rc)
 		}
-		return rc
 	case model.Router:
-		rc = configgen.buildGatewayHTTPRouteConfig(env, node, push, routeName)
-		if rc != nil {
-			rc = envoyfilter.ApplyRouteConfigurationPatches(networking.EnvoyFilter_GATEWAY, node, push, rc)
+		for _, routeName := range routeNames {
```

Comment:
Don't we need similar caching mechanism for gateways? I think the benefit of that would not be significant as sidecars but just throwing it out

Reply:
given how close we are to the 1.3 release (9/10 Tuesday), we should not let the great be the enemy of the good and defer anything non-essential to a subsequent PR.

My understand is protocol sniffing is not in effect for ingress gateway so we don't have to reduce protocol sniffing overhead for ingress gateway.

Reply:
Caching won’t help for gateways as this is list of all routes per port. Gateways are already consolidated per port and the routes for https are very different.

Reply:
Thanks. Makes sense.